Fix E2E tests

gigitux opened the pull request:
This PR improves the stability of our E2E tests.

In this PR:
- some E2E tests were updated to improve the flaky status.
- increased the timeout of `jest-puppeteer`.
- upgraded `@wordpress/e2e-tests"` and `jest-circus` packages
- Disabled `exitOnPageError` option.

####  Disabled `exitOnPageError` option.

This option was disabled to avoid the error:

```
Expected an Error, but "" was thrown. 
```

Sometimes the `JQuery minified file` (from WordPress) throws an error. By default, Puppeteer exits if a script throws an error. `exitOnPageError` option disabled this behavior. 


### Testing
#### User Facing Testing

Run the following command and be sure that there aren't any errors from the E2E folder.

```
npm run test:e2e
```


* [x] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

senadir committed:
don't splitChunk frontend scripts

senadir committed:
simplfy output folders

senadir committed:
reduce minimum chunk size to 10kb

senadir committed:
add all inner blocks for loading

senadir committed:
better splitChunks

senadir committed:
update translation code

senadir committed:
wait for scripts

senadir committed:
try sleeping

senadir committed:
fix loading translation

senadir committed:
fix tests for mini cart

senadir committed:
simplfy po logic

gigitux committed:
Fix E2E tests #6415

Fix E2E tests

opr committed:
Add full state in customer data and exclude it from details check

gigitux committed:
update webpack configuration

opr committed:
Revert change to combobox index

opr committed:
Merge branch 'fix/e2e-fail-translations' of github.com:woocommerce/woocommerce-gutenberg-products-block into fix/e2e-fail-translations

senadir committed:
automate loading chunks

senadir committed:
rename webpack group file

gigitux committed:
Merge branch 'fix/tranlation-not-loading' of github.com:woocommerce/woocommerce-gutenberg-products-block into fix/e2e-fail-translations

gigitux committed:
Merge branch 'fix/e2e-fail-translations' of github.com:woocommerce/woocommerce-gutenberg-products-block into fix/e2e-fail-translations

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-gutenberg-products-block into fix/e2e-fail-translations

senadir committed:
rename block.json path

gigitux committed:
Merge branch 'fix/tranlation-not-loading' of github.com:woocommerce/woocommerce-gutenberg-products-block into fix/e2e-fail-translations

gigitux committed:
Merge branch 'trunk' into fix/e2e-fail-translations

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-gutenberg-products-block into fix/e2e-fail-translations

gigitux committed:
Merge branch 'fix/e2e-fail-translations' of github.com:woocommerce/woocommerce-gutenberg-products-block into fix/e2e-fail-translations

gigitux committed:
remove test

senadir committed:
revert changes to webpack file

senadir committed:
fix loading of chunks

senadir committed:
revert changes to assets api as well

gigitux committed:
Merge branch 'trunk' into fix/e2e-fail-translations

gigitux committed:
Merge branch 'revert/webpack-changes' of github.com:woocommerce/woocommerce-gutenberg-products-block into fix/e2e-fail-translations

gigitux committed:
Merge branch 'fix/e2e-fail-translations' of github.com:woocommerce/woocommerce-gutenberg-products-block into fix/e2e-fail-translations

gigitux committed:
restore wp-env json configuration

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-gutenberg-products-block into fix/e2e-fail-translations

gigitux committed:
restore .wp-env.json

sunyatasattva commented on tests/e2e/config/environment.js:
```diff
@@ -23,4 +23,36 @@ class E2EEnvironment extends PuppeteerEnvironment {
 	}
 }
 
+// This code is useful for debug purposes.
```
I assume you left this comments in on purpose, so that people could easily uncomment it to debug stuff, right?

If that's the case, could you write a bit more on the comment about why and how it is useful and how and when should people use this code?

sunyatasattva commented on tests/e2e/config/jest.setup.js:
```diff
@@ -8,10 +9,17 @@ import {
+ * Internal dependencies
+ */
+import { DEFAULT_TIMEOUT } from '../utils';
+const setDefaultOptions = require( 'expect-puppeteer' ).setDefaultOptions;
```
Sorry, maybe that's a stupid question: can't it be imported normally? Why does it need to be `require`d? Maybe since this is odd, let's put a comment on it?
Can we at least destructure it, you think?

gigitux replied:
Not a stupid question :D 

I copied and paste the import, I will update it 👍

sunyatasattva commented on tests/e2e/specs/shopper/active-filters.test.ts:
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable import/no-extraneous-dependencies */
```
Why is this needed? Because Puppeteer isn't in our `devDependencies`?

gigitux replied:
Puppeteer isn't in our devDependencies because it is installed by `wordpress/e2e-tests` package.

sunyatasattva replied:
Mm, I think that as long as Puppeteer is a peer dependency of another package, that's fine. But since we are now importing something from it, I believe this should be added to the `devDependencies`, what do you think?

Like, if `wordpress/e2e-tests` dependencies change (even in version) this might cause errors in this code. Since we are importing directly, I think it now becomes an internal dep.

sunyatasattva commented on tests/e2e/specs/shopper/active-filters.test.ts:
```diff
@@ -36,6 +34,8 @@ const block = {
 		frontend: {
 			activeFilterType: '.wc-block-active-filters__list-item-type',
 			activeFilterName: '.wc-block-active-filters__list-item-name',
+			inStockInput: "label[for='instock']",
+			'128gbInput': "label[for='128gb']",
```
I don't see those selectors used anywhere, maybe you didn't need them anymore?

As a suggestion, maybe let's use a selector factory here, something like:

```suggestion
			filterCheckboxFor( label: string ) => `label[for="${ label }"]`
```

gigitux replied:
Removed the not used selectors 👍

sunyatasattva commented on tests/e2e/specs/shopper/active-filters.test.ts:
```diff
@@ -113,18 +113,21 @@ describe( 'Shopper → Active Filters Block', () => {
 			);
 
+			await page.waitForSelector(
+				selectors.frontend.productsList + '.is-loading',
```
Just curious, why did you use this method of concatenation over template literal?

gigitux replied:
Just for velocity. I will update the code

sunyatasattva commented on tests/e2e/specs/shopper/active-filters.test.ts:
```diff
@@ -113,18 +113,21 @@ describe( 'Shopper → Active Filters Block', () => {
 			} );
 
-			await page.waitForSelector( block.class );
+			await page.screenshot( {
+				path: 'aaa.png',
+				fullPage: true,
+			} );
 
```
This has probably been forgotten, I assume? 🤔

gigitux replied:
ahahah, yes! Powerful debugger :P

sunyatasattva commented on tests/e2e/specs/shopper/active-filters.test.ts:
```diff
@@ -113,18 +113,21 @@ describe( 'Shopper → Active Filters Block', () => {
-					text: 'Stock Status',
-				}
-			);
+			await expect( page ).toMatch( 'Stock Status' );
```
I like that you have refactored the result expectation out (although I probably wouldn't have put it as a property of the `block` object, as it seems a bit confusing), what do you think about getting this string out as well? I'm not sure what it is, but something like `FILTER_TITLE` ?

gigitux replied:
Thanks for pointing this out. [I extracted more string](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/42e552d5f28c64f6e675f5a0a3eb145d934a3088/tests/e2e/specs/shopper/active-filters.test.ts/#L51-L54). Do you think that we should put these constants in `utils/constants.js`. I think that for these one it doesn't make sense because they are not for general purpose, but just for this block.

sunyatasattva replied:
I agree with you. If they are only for this block, it's ok to have them scoped in this file. If you can see some other test reusing them, it'd be nice to have them in the constants file.

sunyatasattva commented on tests/e2e/specs/shopper/active-filters.test.ts:
```diff
@@ -144,7 +147,7 @@ describe( 'Shopper → Active Filters Block', () => {
 		} );
 
-		fit( 'When clicking the X on a filter it removes a filter', async () => {
+		it( 'When clicking the X on a filter it removes a filter', async () => {
```
Hehe fit! 💪

sunyatasattva commented on tests/e2e/specs/shopper/cart-checkout/account.test.js:
```diff
@@ -15,7 +15,7 @@ import { visitAdminPage } from '@wordpress/e2e-test-utils';
  */
-import { shopper, merchant } from '../../../../utils';
+import { shopper, merchant, clickLink } from '../../../../utils';
 import { SIMPLE_PHYSICAL_PRODUCT_NAME } from '.../../../../utils/constants';
```
Just noticed this!

Regarding my comment above, this is a much better way to put the string instead of inside the `block` object IMHO. Also, it's literally the same string as your `foundProduct`, so what do you say about using this constant instead of the `foundProduct` above?

sunyatasattva commented on tests/e2e/specs/shopper/filter-products-by-price.test.ts:
```diff
@@ -59,7 +59,7 @@ const setMaxPrice = async () => {
 		clickCount: 3,
 	} );
-	await page.keyboard.type( '1.99' );
+	await page.keyboard.type( '2' );
```
Haha that's interesting. Why this change?

gigitux replied:
Just for debug reason, I restored the change.

sunyatasattva commented on tests/e2e/specs/shopper/filter-products-by-price.test.ts:
```diff
@@ -97,10 +97,10 @@ describe( `${ block.name } Block`, () => {
 			await page.waitForNetworkIdle();
 			await waitForAllProductsBlockLoaded();
-			const products = await page.$$( selectors.frontend.productsList );
+
+			await page.waitForTimeout( 1000 );
```
Rogue hard coded timeout, probably forgotten?

gigitux replied:
Removed 👍

sunyatasattva commented on tests/e2e/specs/shopper/filter-products-by-price.test.ts:
```diff
@@ -196,14 +191,10 @@ describe( `${ block.name } Block`, () => {
 			const parsedURL = new URL( pageURL );
 
 			expect( isRefreshed ).toBeCalledTimes( 1 );
-			expect( products ).toHaveLength( 1 );
```
Why are we removing this check?

gigitux replied:
Re-added 👍

sunyatasattva reviewed (changes requested):
Added a few comments. Code looks good!

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-gutenberg-products-block into fix/e2e-fail-translations

gigitux committed:
address feedback

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-gutenberg-products-block into fix/e2e-fail-translations

gigitux committed:
add comment

gigitux committed:
address comments

gigitux committed:
improve utils function

gigitux commented:
> Added a few comments. Code looks good!

Thanks for your great review! I address your feedback. I extract[ `waitForAllProductsBlockLoaded` function]( https://github.com/woocommerce/woocommerce-blocks/blob/9895da816b3fb04ed0b567c48df3c03689c5b433/tests/e2e/utils.js#L451-L456) and moved in the utils file.

I guess that we should start to split the utils.js file. It has become huge.

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-gutenberg-products-block into fix/e2e-fail-translations

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-gutenberg-products-block into fix/e2e-fail-translations

sunyatasattva commented on tests/e2e/utils.js:
```diff
@@ -448,14 +448,10 @@ export const openBlockEditorSettings = async () => {
-		page.click( selector ),
-		page.waitForNavigation( { waitUntil: 'networkidle0' } ),
-	] );
+export const waitForAllProductsBlockLoaded = async () => {
```
Small nitpick, not blocking:

1. Could you add a little JSDoc to this? I know the name is self explanatory, but it might be worth explaining what it does and it also encourages contributors to always put JSDocs on their functions.
2. I know a lot of selectors here are hard-coded, but I would like to see it extracted somewhere else, if you have the patience for that 😊

sunyatasattva reviewed (approved):
I'm pre-approving, but I'd like to see Puppeteer being added to `devDependencies`. The other change is optional.

Great job and thanks for addressing all my issues and creating clean code!

I agree we should start thinking of splitting our utils file!

gigitux committed:
install puppeteer

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-gutenberg-products-block into fix/e2e-fail-translations

gigitux commented:
> I'm pre-approving, but I'd like to see Puppeteer being added to `devDependencies`. The other change is optional.
> 
> Great job and thanks for addressing all my issues and creating clean code!
> 
> I agree we should start thinking of splitting our utils file!

I addressed all your suggestions! Thanks again! 🎉
